README.md
[![Build Status](https://travis-ci.org/fengari-lua/fengari-web.svg?branch=master)](https://travis-ci.org/fengari-lua/fengari-web)
[![npm](https://img.shields.io/npm/v/fengari-web.svg)](https://npmjs.com/package/fengari-web)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![#fengari on Freenode](https://img.shields.io/Freenode/%23fengari.png)](https://webchat.freenode.net/?channels=fengari)


# fengari-web

Provides anything you need to run [Fengari](https://fengari.io) in the browser.

## Getting started

### Download

Visit the [GitHub releases page](https://github.com/fengari-lua/fengari-web/releases) and get the latest version.

Alternatively you can [Build](#Building) fengari-web yourself.


### Directly

Load fengari-web in your web page:

```html
<script src="fengari-web.js" type="text/javascript"></script>
```

Now any script of type `application/lua` will be run by fengari:

```html
<script type="application/lua">
print("Hello World!")
</script>

<script src="/my-script.lua" type="application/lua" async></script>
```

Note that if you use a `src` attribute, it is strongly recommended for it to be [`async`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-async).


### With build process

See [fengari-loader](https://github.com/fengari-lua/fengari-loader/)


## Compatibility

fengari-web should work in all modern browsers.

Verified to work in:

  - Chrome >= 38
  - Firefox >= 19
  - Safari >= 8
  - Microsoft IE 11
  - Microsoft Edge


## API

As well as running `<script type="application/lua">` tags, fengari-web creates a `fengari` global that contains the [core `fengari` API](https://github.com/fengari-lua/fengari#the-js-api) supplemented with:

  - `L`: the main `lua_State` (in which script tags are run)
  - `interop`: containing the [fengari-interop](https://github.com/fengari-lua/fengari-interop) library
  - `load(source, chunkname)`: a function that loads the lua code in `source` with the optional chunk name `chunkname` and returns it as a function.
    This function can be used to programmatically run lua code in the main `lua_State` from JavaScript. e.g.
    ```js
    console.log(fengari.load('return 1+1')())
    ```


## Building

```bash
git clone https://github.com/fengari-lua/fengari-web.git
npm install
```

This should automatically kick off the build process.
The built files can then be found in the `dist/` directory.

If you need to rebuild, run

```bash
npm run build
```

Or [use `webpack` directly](https://webpack.js.org/api/cli/).


NEWS
0.1.4 - 2018-11-19

  - Release is to fix broken npm package


0.1.3 - 2018-11-11

  - Fix issues building with newer webpack/babel


0.1.2 - 2018-04-23

  - Change check for non-webworker to a check for document object


0.1.1 - 2018-04-03

  - Fix CSS selector so that script tags present on fengari-web load are run


0.1.0 - 2018-04-01

  - Runs `<script type="application/lua">` tags from HTML
  - `fengari.load` to compile lua source code into a JavaScript function


lua-web-cli.js
/*jshint esversion: 6 */
"use strict";

window.WEB = true;

const lua          = fengari.lua;
const lauxlib      = fengari.lauxlib;
const lualib       = fengari.lualib;

const stdin = lua.to_luastring("=stdin");
const _PROMPT = lua.to_luastring("_PROMPT");
const _PROMPT2 = lua.to_luastring("_PROMPT2");

const output = document.getElementById('fengari-console');
const input = document.getElementById('fengari-input');
const prompt = document.getElementById('fengari-prompt');


const triggerEvent = function(el, type) {
    var e = document.createEvent('HTMLEvents');
    e.initEvent(type, false, true);
    el.dispatchEvent(e);
}

const out = function(msg) {
    output.innerHTML += msg;
    output.scrollTop = output.scrollHeight;
};

const luaW_print = function(L) {
    let n = lua.lua_gettop(L); /* number of arguments */
    let str = [];

    lua.lua_getglobal(L, lua.to_luastring("tostring", true));
    for (let i = 1; i <= n; i++) {
        lua.lua_pushvalue(L, -1);  /* function to be called */
        lua.lua_pushvalue(L, i);  /* value to print */
        lua.lua_call(L, 1, 1);
        let s = lua.lua_tolstring(L, -1);
        if (s === null)
            return lauxlib.luaL_error(L, lua.to_luastring("'tostring' must return a string to 'print'", true));
        if (i > 1) s = ["\t".charCodeAt(0)].concat(s);
        str = str.concat(s);
        lua.lua_pop(L, 1);
    }

    out(lua.to_jsstring(str) + "\n");
    return 0;
};

const report = function(L, status) {
    if (status !== lua.LUA_OK) {
        out(`${lua.lua_tojsstring(L, -1)}\n`);
        lua.lua_pop(L, 1);
    }
    return status;
};

const msghandler = function(L) {
    let msg = lua.lua_tostring(L, 1);
    if (msg === null) {  /* is error object not a string? */
        if (lauxlib.luaL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
          lua.lua_type(L, -1) == lua.LUA_TSTRING)  /* that produces a string? */
            return 1;  /* that is the message */
        else
            msg = lua.lua_pushstring(L, lua.to_luastring(`(error object is a ${lua.to_jsstring(lauxlib.luaL_typename(L, 1))} value)`));
    }
    lauxlib.luaL_traceback(L, L, msg, 1);  /* append a standard traceback */

    out(lua.lua_tojsstring(L, -1));

    return 1;  /* return the traceback */
};

const docall = function(L, narg, nres) {
    let base = lua.lua_gettop(L) - narg;
    lua.lua_pushcfunction(L, msghandler);
    lua.lua_insert(L, base);
    let status = lua.lua_pcall(L, narg, nres, base);
    lua.lua_remove(L, base);
    return status;
};

const doREPL = function(L) {
    lua.lua_getglobal(L, _PROMPT);
    prompt.innerHTML = ">";
    out("\n> " + input.value + "\n");
    lua.lua_pop(L, 1);

    if (input.value.length === 0)
        return;

    let status;
    {
        let buffer = lua.to_luastring("return " + input.value);
        status = lauxlib.luaL_loadbuffer(L, buffer, buffer.length, stdin);
    }
    if (status !== lua.LUA_OK) {
        lua.lua_pop(L, 1);
        let buffer = lua.to_luastring(input.value);
        if (lauxlib.luaL_loadbuffer(L, buffer, buffer.length, stdin) === lua.LUA_OK) {
            status = lua.LUA_OK;
        }
    }
    if (status === lua.LUA_OK) {
        status = docall(L, 0, lua.LUA_MULTRET);
    }
    if (status === lua.LUA_OK) {
        let n = lua.lua_gettop(L);
        if (n > 0) {  /* any result to be printed? */
            luaW_print(L);
        }
    } else {
        report(L, status);
    }
    lua.lua_settop(L, 0);  /* remove eventual returns */

    input.value = "";

    triggerEvent(output, 'change');
};

const L = lauxlib.luaL_newstate();

output.innerHTML += lua.FENGARI_COPYRIGHT + "\n";
/* open standard libraries */
lualib.luaL_openlibs(L);

// Overwrite luaB_print
lua.lua_register(L, lua.to_luastring("print"), luaW_print);

input.onkeypress = function(e){
    if (!e) e = window.event;
    var keyCode = e.keyCode || e.which;
    if (keyCode == '13'){
        doREPL(L);
        return false;
    }
};


src/web-cli.js
/*jshint esversion: 6 */
"use strict";

window.WEB = true;

const fengari  = require('fengari');
const lua      = fengari.lua;
const lauxlib  = fengari.lauxlib;
const lualib   = fengari.lualib;
const interop  = require('fengari-interop');


const stdin = lua.to_luastring("=stdin");

const output = document.getElementById('fengari-console');
const input = document.getElementById('fengari-input');


const triggerEvent = function(el, type) {
    var e = document.createEvent('HTMLEvents');
    e.initEvent(type, false, true);
    el.dispatchEvent(e);
};

const out = function(msg) {
    output.textContent += msg;
    output.scrollTop = output.scrollHeight;
};

const luaW_print = function(L) {
    let n = lua.lua_gettop(L); /* number of arguments */
    let str = [];

    lua.lua_getglobal(L, lua.to_luastring("tostring", true));
    for (let i = 1; i <= n; i++) {
        lua.lua_pushvalue(L, -1);  /* function to be called */
        lua.lua_pushvalue(L, i);  /* value to print */
        lua.lua_call(L, 1, 1);
        let s = lua.lua_tolstring(L, -1);
        if (s === null)
            return lauxlib.luaL_error(L, lua.to_luastring("'tostring' must return a string to 'print'", true));
        if (i > 1) s = ["\t".charCodeAt(0)].concat(s);
        str = str.concat(s);
        lua.lua_pop(L, 1);
    }

    out(lua.to_jsstring(str) + "\n");
    return 0;
};

const report = function(L, status) {
    if (status !== lua.LUA_OK) {
        out(`${lua.lua_tojsstring(L, -1)}\n`);
        lua.lua_pop(L, 1);
    }
    return status;
};

const msghandler = function(L) {
    let msg = lua.lua_tostring(L, 1);
    if (msg === null) {  /* is error object not a string? */
        if (lauxlib.luaL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
          lua.lua_type(L, -1) == lua.LUA_TSTRING)  /* that produces a string? */
            return 1;  /* that is the message */
        else
            msg = lua.lua_pushstring(L, lua.to_luastring(`(error object is a ${lua.to_jsstring(lauxlib.luaL_typename(L, 1))} value)`));
    }
    lauxlib.luaL_traceback(L, L, msg, 1);  /* append a standard traceback */

    out(lua.lua_tojsstring(L, -1));

    return 1;  /* return the traceback */
};

const docall = function(L, narg, nres) {
    let base = lua.lua_gettop(L) - narg;
    lua.lua_pushcfunction(L, msghandler);
    lua.lua_insert(L, base);
    let status = lua.lua_pcall(L, narg, nres, base);
    lua.lua_remove(L, base);
    return status;
};

const doREPL = function(L) {
    out("\n> " + input.value + "\n");

    if (input.value.length === 0)
        return;

    let status;
    {
        let buffer = lua.to_luastring("return " + input.value);
        status = lauxlib.luaL_loadbuffer(L, buffer, buffer.length, stdin);
    }
    if (status !== lua.LUA_OK) {
        lua.lua_pop(L, 1);
        let buffer = lua.to_luastring(input.value);
        if (lauxlib.luaL_loadbuffer(L, buffer, buffer.length, stdin) === lua.LUA_OK) {
            status = lua.LUA_OK;
        }
    }
    if (status === lua.LUA_OK) {
        status = docall(L, 0, lua.LUA_MULTRET);
    }
    if (status === lua.LUA_OK) {
        let n = lua.lua_gettop(L);
        if (n > 0) {  /* any result to be printed? */
            luaW_print(L);
        }
    } else {
        report(L, status);
    }
    lua.lua_settop(L, 0);  /* remove eventual returns */

    input.value = "";

    triggerEvent(output, 'change');
};

const L = lauxlib.luaL_newstate();

output.textContent += lua.FENGARI_COPYRIGHT + "\n";
/* open standard libraries */
lualib.luaL_openlibs(L);

// interop.luaopen_js(L);

lauxlib.luaL_requiref(L, lua.to_luastring("js"), interop.luaopen_js, 1);
lua.lua_pop(L, 1); /* remove lib */

// Overwrite luaB_print
lua.lua_register(L, lua.to_luastring("print"), luaW_print);

input.onkeypress = function(e){
    if (!e) e = window.event;
    var keyCode = e.keyCode || e.which;
    if (keyCode == '13' && !!!e.shiftKey) {
        doREPL(L);
        return false;
    }
};


src/web-cli.lua
local js = require "js"

-- Save references to lua baselib functions used
local _G = _G
local load = load
local pack, unpack = table.pack, table.unpack
local tostring = tostring
local traceback = debug.traceback
local xpcall = xpcall

local document = js.global.document
local hljs = js.global.hljs
local rangy = js.global.rangy

local output = document:getElementById("fengari-console")
local prompt = document:getElementById("fengari-prompt")
local input = document:getElementById("fengari-input")
assert(output and prompt and input)

local function triggerEvent(el, type)
    local e = document:createEvent("HTMLEvents")
    e:initEvent(type, false, true)
    el:dispatchEvent(e)
end

local history = {}
local historyIndex = nil
local historyLimit = 100

_G.print = function(...)
    local toprint = pack(...)

    local line = document:createElement("span")
    output:appendChild(line)

    for i = 1, toprint.n do
        local item = document:createElement("pre")
        item.style["white-space"] = "pre-wrap"
        item:appendChild(document:createTextNode(tostring(toprint[i])))
        line:appendChild(item)
    end

    output.scrollTop = output.scrollHeight
end

local function doREPL()
    do
        local line = document:createElement("span")
        line:appendChild(document:createTextNode(prompt.textContent))
        local item = document:createElement("pre")
        item.className = "lua"
        item.style.padding = "0"
        item.style.display = "inline"
        item.style["white-space"] = "pre-wrap"
        item.textContent = input.textContent
        hljs:highlightBlock(item)
        line:appendChild(item)
        output:appendChild(line)
        output:appendChild(document:createElement("br"))
        output.scrollTop = output.scrollHeight
    end

    if input.textContent.length == 0 then
        return
    end

    local line = input.textContent
    table.insert(history, line)
    if #history > historyLimit then
        table.remove(history, 1)
    end

    local fn, err = load("return " .. line, "stdin")
    if not fn then
        fn, err = load(line, "stdin")
    end

    if fn then
        local results = pack(xpcall(fn, traceback))
        if results[1] then
            if results.n > 1 then
                _G.print(unpack(results, 2, results.n))
            end
        else
            _G.print(results[2])
        end
    else
        _G.print(err)
    end

    input.textContent = ""
    prompt.textContent = _G._PROMPT or "> "

    triggerEvent(output, "change")
end

function input:onkeydown(e)
    local position = rangy:saveSelection()
    hljs:highlightBlock(input)
    rangy:restoreSelection(position)

    if not e then
        e = js.global.event
    end

    local key = e.key or e.which
    if key == "Enter" and not e.shiftKey then
        doREPL()
        return false
    elseif key == "ArrowUp" then
        historyIndex = historyIndex and historyIndex - 1 or #history
        historyIndex = historyIndex > 0 and historyIndex or 1
        input.textContent = history[historyIndex]
        return false
    elseif key == "ArrowDown" then
        historyIndex = historyIndex and historyIndex + 1
        historyIndex = historyIndex <= #history and historyIndex or #history
        input.textContent = history[historyIndex]
        return false
    end
end

_G.print(_G._COPYRIGHT)


webpack.config.js
"use strict";

const webpack = require('webpack');

module.exports = [
	{
		/*
		This target produces a file suitable for embedding onto any webpage via a <script> tag.
		*/
		name: 'script-tag',
		entry: './src/fengari-web.js',
		target: 'web',
		output: {
			filename: 'fengari-web.js',
			library: 'fengari',
			libraryTarget: 'umd'
		},
		devtool: 'hidden-source-map',
		node: false,
		module: {
			rules: [
				{
					test: [/\.js$/],
					loader: 'babel-loader',
					options: {
						presets: [['@babel/preset-env', {
							"targets": {
								"browsers": ["last 2 versions", "not safari <= 7", "not ie <= 10"]
							}
						}]]
					}
				}
			]
		},
		plugins: [
			new webpack.DefinePlugin({
				"process.env.FENGARICONF": "void 0",
				"typeof process": JSON.stringify("undefined")
			})
		]
	},
	{
		/*
		This target exists to create a bundle that has the node-specific paths eliminated.
		It is expected that most people would minify this with their own build process
		*/
		name: 'bundle',
		entry: './src/fengari-web.js',
		target: 'web',
		output: {
			filename: 'fengari-web.bundle.js',
			libraryTarget: 'commonjs2'
		},
		devtool: 'hidden-source-map',
		node: false,
		plugins: [
			new webpack.DefinePlugin({
				"process.env.FENGARICONF": "void 0",
				"typeof process": JSON.stringify("undefined")
			})
		]
	}
];


src/web-cli-lua.js
"use strict";

const fengari  = require('fengari');
const lua      = fengari.lua;
const lauxlib  = fengari.lauxlib;
const lualib   = fengari.lualib;
const interop  = require('fengari-interop');

const L = lauxlib.luaL_newstate();

/* open standard libraries */
lualib.luaL_openlibs(L);
lauxlib.luaL_requiref(L, lua.to_luastring("js"), interop.luaopen_js, 1);
lua.lua_pop(L, 1); /* remove lib */

lua.lua_pushstring(L, lua.to_luastring(lua.FENGARI_COPYRIGHT));
lua.lua_setglobal(L, lua.to_luastring("_COPYRIGHT"));

const msghandler = function(L) {
	let ar = new lua.lua_Debug();
	lua.lua_getstack(L, 2, ar);
	lua.lua_getinfo(L, lua.to_luastring("Sl"), ar);
	interop.push(L, new ErrorEvent("error", {
		bubbles: true,
		cancelable: true,
		message: lua.lua_tojsstring(L, 1),
		error: interop.tojs(L, 1),
		filename: lua.to_jsstring(ar.short_src),
		lineno: ar.currentline > 0 ? ar.currentline : void 0
	}));
	return 1;
}

const run_lua_script = function(tag, code, chunkname) {
	let ok = lauxlib.luaL_loadbuffer(L, code, null, chunkname);
	let e;
	if (ok === lua.LUA_ERRSYNTAX) {
		let msg = lua.lua_tojsstring(L, -1);
		let filename = tag.src?tag.src:document.location;
		let lineno = void 0; /* TODO: extract out of msg */
		let syntaxerror = new SyntaxError(msg, filename, lineno);
		e = new ErrorEvent("error", {
			message: msg,
			error: syntaxerror,
			filename: filename,
			lineno: lineno
		});
	} else if (ok === lua.LUA_OK) {
		/* insert message handler below function */
		let base = lua.lua_gettop(L);
		lua.lua_pushcfunction(L, msghandler);
		lua.lua_insert(L, base);
		/* set document.currentScript.
		   We can't set it normally; but we can create a getter for it, then remove the getter */
		Object.defineProperty(document, 'currentScript', {
			value: tag,
			configurable: true
		});
		ok = lua.lua_pcall(L, 0, 0, base);
		/* Remove the currentScript getter installed above; this restores normal behaviour */
		delete document.currentScript;
		/* Check if normal error that msghandler would have handled */
		if (ok === lua.LUA_ERRRUN) {
			e = interop.checkjs(L, -1);
		}
	}
	if (ok !== lua.LUA_OK) {
		if (e === void 0) {
			e = new ErrorEvent("error", {
				message: lua.lua_tojstring(L, -1),
				error: interop.tojs(L, -1)
			});
		}
		lua.lua_pop(L, 1);
		if (window.dispatchEvent(e)) {
			console.error("uncaught exception", e.error);
		}
	}
};

const crossorigin_to_credentials = function(crossorigin) {
	switch(crossorigin) {
		case "anonymous": return "omit";
		case "use-credentials": return "include";
		default: return "same-origin";
	}
};

const run_lua_script_tag = function(tag) {
	if (tag.src) {
		let chunkname = lua.to_luastring("@"+tag.src);
		/* JS script tags are async after document has loaded */
		if (document.readyState === "complete" || tag.async) {
			fetch(tag.src, {
				method: "GET",
				credentials: crossorigin_to_credentials(tag.crossorigin),
				redirect: "follow",
				integrity: tag.integrity
			}).then(function(resp) {
				if (resp.ok) {
					resp.arrayBuffer().then(function(buffer) {
						let code = Array.from(new Uint8Array(buffer));
						run_lua_script(tag, code, chunkname);
					});
				} else {
					tag.dispatchEvent(new Event("error"));
				}
			});
		} else {
			/* Needs to be synchronous: use an XHR */
			let xhr = new XMLHttpRequest();
			xhr.open("GET", tag.src, false);
			xhr.send();
			if (xhr.status >= 200 && xhr.status < 300) {
				/* TODO: subresource integrity check? */
				let code = lua.to_luastring(xhr.response);
				run_lua_script(tag, code, chunkname);
			} else {
				tag.dispatchEvent(new Event("error"));
			}
		}
	} else {
		let code = lua.to_luastring(tag.innerHTML);
		let chunkname = tag.id ? lua.to_luastring("="+tag.id) : code;
		run_lua_script(tag, code, chunkname);
	}
};

/* watch for new <script type="text/lua"> tags added to document */
(new MutationObserver(function(records, observer) {
    for (let r=0; r<records.length; r++) {
        for (let i=0; i<records[r].addedNodes.length; i++) {
            let tag = records[r].addedNodes[i];
            if (tag.tagName == "SCRIPT" && tag.type == "text/lua") {
                run_lua_script_tag(tag);
            }
        }
    }
})).observe(document, {
    childList: true,
    subtree: true
});

/* run existing <script type="text/lua"> tags */
Array.prototype.forEach.call(document.querySelectorAll('script[type=\"text\/lua\"]'), run_lua_script_tag);


src/fengari-web.js
"use strict";

import {
	FENGARI_AUTHORS,
	FENGARI_COPYRIGHT,
	FENGARI_RELEASE,
	FENGARI_VERSION,
	FENGARI_VERSION_MAJOR,
	FENGARI_VERSION_MINOR,
	FENGARI_VERSION_NUM,
	FENGARI_VERSION_RELEASE,

	luastring_eq,
	luastring_indexOf,
	luastring_of,
	to_jsstring,
	to_luastring,
	to_uristring,

	lua,
	lauxlib,
	lualib
} from 'fengari';
import * as interop from 'fengari-interop';

const {
	LUA_ERRRUN,
	LUA_ERRSYNTAX,
	LUA_OK,
	LUA_VERSION_MAJOR,
	LUA_VERSION_MINOR,
	lua_Debug,
	lua_getinfo,
	lua_getstack,
	lua_gettop,
	lua_insert,
	lua_pcall,
	lua_pop,
	lua_pushcfunction,
	lua_pushstring,
	lua_remove,
	lua_setglobal,
	lua_tojsstring
} = lua;
const {
	luaL_loadbuffer,
	luaL_newstate,
	luaL_requiref
} = lauxlib;
const {
	checkjs,
	luaopen_js,
	push,
	tojs
} = interop;

export {
	FENGARI_AUTHORS,
	FENGARI_COPYRIGHT,
	FENGARI_RELEASE,
	FENGARI_VERSION,
	FENGARI_VERSION_MAJOR,
	FENGARI_VERSION_MINOR,
	FENGARI_VERSION_NUM,
	FENGARI_VERSION_RELEASE,

	luastring_eq,
	luastring_indexOf,
	luastring_of,
	to_jsstring,
	to_luastring,
	to_uristring,

	lua,
	lauxlib,
	lualib,
	interop
};

export const L = luaL_newstate();

/* open standard libraries */
lualib.luaL_openlibs(L);
luaL_requiref(L, to_luastring("js"), luaopen_js, 1);
lua_pop(L, 1); /* remove lib */

lua_pushstring(L, to_luastring(FENGARI_COPYRIGHT));
lua_setglobal(L, to_luastring("_COPYRIGHT"));

/* Helper function to load a JS string of Lua source */
export function load(source, chunkname) {
	if (typeof source == "string")
		source = to_luastring(source);
	else if (!(source instanceof Uint8Array))
		throw new TypeError("expects an array of bytes or javascript string");

	chunkname = chunkname?to_luastring(chunkname):null;
	let ok = luaL_loadbuffer(L, source, null, chunkname);
	let res;
	if (ok === LUA_ERRSYNTAX) {
		res = new SyntaxError(lua_tojsstring(L, -1));
	} else {
		res = tojs(L, -1);
	}
	lua_pop(L, 1);
	if (ok !== LUA_OK) {
		throw res;
	}
	return res;
}

if (typeof document !== 'undefined' && document instanceof HTMLDocument) {
	/* Have a document, e.g. we are in main browser window */

	const crossorigin_to_credentials = function(crossorigin) {
		switch(crossorigin) {
			case "anonymous": return "omit";
			case "use-credentials": return "include";
			default: return "same-origin";
		}
	};

	const msghandler = function(L) {
		let ar = new lua_Debug();
		if (lua_getstack(L, 2, ar))
			lua_getinfo(L, to_luastring("Sl"), ar);
		push(L, new ErrorEvent("error", {
			bubbles: true,
			cancelable: true,
			message: lua_tojsstring(L, 1),
			error: tojs(L, 1),
			filename: ar.short_src ? to_jsstring(ar.short_src) : void 0,
			lineno: ar.currentline > 0 ? ar.currentline : void 0
		}));
		return 1;
	};

	const run_lua_script = function(tag, code, chunkname) {
		let ok = luaL_loadbuffer(L, code, null, chunkname);
		let e;
		if (ok === LUA_ERRSYNTAX) {
			let msg = lua_tojsstring(L, -1);
			let filename = tag.src?tag.src:document.location;
			let lineno = void 0; /* TODO: extract out of msg */
			let syntaxerror = new SyntaxError(msg, filename, lineno);
			e = new ErrorEvent("error", {
				message: msg,
				error: syntaxerror,
				filename: filename,
				lineno: lineno
			});
		} else if (ok === LUA_OK) {
			/* insert message handler below function */
			let base = lua_gettop(L);
			lua_pushcfunction(L, msghandler);
			lua_insert(L, base);
			/* set document.currentScript.
			   We can't set it normally; but we can create a getter for it, then remove the getter */
			Object.defineProperty(document, 'currentScript', {
				value: tag,
				configurable: true
			});
			ok = lua_pcall(L, 0, 0, base);
			/* Remove the currentScript getter installed above; this restores normal behaviour */
			delete document.currentScript;
			/* Remove message handler */
			lua_remove(L, base);
			/* Check if normal error that msghandler would have handled */
			if (ok === LUA_ERRRUN) {
				e = checkjs(L, -1);
			}
		}
		if (ok !== LUA_OK) {
			if (e === void 0) {
				e = new ErrorEvent("error", {
					message: lua_tojsstring(L, -1),
					error: tojs(L, -1)
				});
			}
			lua_pop(L, 1);
			if (window.dispatchEvent(e)) {
				console.error("uncaught exception", e.error);
			}
		}
	};

	const process_xhr_response = function(xhr, tag, chunkname) {
		if (xhr.status >= 200 && xhr.status < 300) {
			let code = xhr.response;
			if (typeof code === "string") {
				code = to_luastring(xhr.response);
			} else { /* is an array buffer */
				code = new Uint8Array(code);
			}
			/* TODO: subresource integrity check? */
			run_lua_script(tag, code, chunkname);
		} else {
			tag.dispatchEvent(new Event("error"));
		}
	};

	const run_lua_script_tag = function(tag) {
		if (tag.src) {
			let chunkname = to_luastring("@"+tag.src);
			/* JS script tags are async after document has loaded */
			if (document.readyState === "complete" || tag.async) {
				if (typeof fetch === "function") {
					fetch(tag.src, {
						method: "GET",
						credentials: crossorigin_to_credentials(tag.crossorigin),
						redirect: "follow",
						integrity: tag.integrity
					}).then(function(resp) {
						if (resp.ok) {
							return resp.arrayBuffer();
						} else {
							throw new Error("unable to fetch");
						}
					}).then(function(buffer) {
						let code = new Uint8Array(buffer);
						run_lua_script(tag, code, chunkname);
					}).catch(function(reason) {
						tag.dispatchEvent(new Event("error"));
					});
				} else {
					let xhr = new XMLHttpRequest();
					xhr.open("GET", tag.src, true);
					xhr.responseType = "arraybuffer";
					xhr.onreadystatechange = function() {
						if (xhr.readyState === 4)
							process_xhr_response(xhr, tag, chunkname);
					};
					xhr.send();
				}
			} else {
				/* Needs to be synchronous: use an XHR */
				let xhr = new XMLHttpRequest();
				xhr.open("GET", tag.src, false);
				xhr.send();
				process_xhr_response(xhr, tag, chunkname);
			}
		} else {
			let code = to_luastring(tag.innerHTML);
			let chunkname = tag.id ? to_luastring("="+tag.id) : code;
			run_lua_script(tag, code, chunkname);
		}
	};

	const contentTypeRegexp = /^(.*?\/.*?)([\t ]*;.*)?$/;
	const luaVersionRegex = /^(\d+)\.(\d+)$/;
	const try_tag = function(tag) {
		if (tag.tagName !== "SCRIPT")
			return;

		/* strip off mime type parameters */
		let contentTypeMatch = contentTypeRegexp.exec(tag.type);
		if (!contentTypeMatch)
			return;
		let mimetype = contentTypeMatch[1];
		if (mimetype !== "application/lua" && mimetype !== "text/lua")
			return;

		if (tag.hasAttribute("lua-version")) {
			let lua_version = luaVersionRegex.exec(tag.getAttribute("lua-version"));
			if (!lua_version || lua_version[1] !== LUA_VERSION_MAJOR || lua_version[2] !== LUA_VERSION_MINOR)
				return;
		}

		run_lua_script_tag(tag);
	};

	if (typeof MutationObserver !== 'undefined') {
		/* watch for new script tags added to document */
		(new MutationObserver(function(records, observer) {
			for (let i=0; i<records.length; i++) {
				let record = records[i];
				for (let j=0; j<record.addedNodes.length; j++) {
					try_tag(record.addedNodes[j]);
				}
			}
		})).observe(document, {
			childList: true,
			subtree: true
		});
	} else if (console.warn) {
		console.warn("fengari-web: MutationObserver not found; lua script tags will not be run when inserted");
	}

	/* the query selector here is slightly liberal,
	   more checks occur in try_tag */
	const selector = 'script[type^="application/lua"], script[type^="text/lua"]';

	/* try to run existing script tags */
	Array.prototype.forEach.call(document.querySelectorAll(selector), try_tag);
}
